### ISSUE TITLE
Stateful Model can't run on ANE, CPU & GPU are fine.

### ISSUE THREAD

**Issue by alexchen4ai** [Anemll/Anemll#27]
I am trying to implement a Spanish LLM model with self attention and KV cache, and run it on Apple Neural Engine by referring to the modeling llama.py in this repo, but I found I can't make it run on neural engine.

To help replicate the problem, I have extracted some minimal code here:

```python
import torch
import numpy as np
import torch.nn as nn
import coremltools as ct

SEQ_LEN, CONTEXT_LEN = 128, 1024
head_dim, hidden_size, num_attention_heads = 128, 1024, 16

class SelfAttention(nn.Module):
    def __init__(self, head_dim, hidden_size, num_attention_heads, context_len=1024):
        super().__init__()
        self.head_dim = head_dim

        self.q_proj = nn.Linear(hidden_size, num_attention_heads * head_dim, bias=False)
        self.k_proj = nn.Linear(hidden_size, num_attention_heads * head_dim, bias=False)
        self.v_proj = nn.Linear(hidden_size, num_attention_heads * head_dim, bias=False)
        self.register_buffer("k_cache", torch.zeros(num_attention_heads, context_len, head_dim))
        self.register_buffer("v_cache", torch.zeros(num_attention_heads, context_len, head_dim))

    @torch.no_grad()
    def forward(self, hidden_states, attention_mask):
        seq_len = hidden_states.shape[1]
        hidden_shape = (1, seq_len, -1, self.head_dim)
        query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
        key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
        
        self.k_cache[:, :seq_len, :] = key_states.squeeze(0)
        self.v_cache[:, :seq_len, :] = value_states.squeeze(0)
        curr_key_states = self.k_cache.unsqueeze(0)
        curr_value_states = self.v_cache.unsqueeze(0)

        attn_weights = torch.matmul(query_states, curr_key_states.transpose(2, 3))
        causal_mask = attention_mask[:, :, :, : curr_key_states.shape[-2]]
        attn_weights = attn_weights + causal_mask

        attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
        attn_output = torch.matmul(attn_weights, curr_value_states)
        attn_output = attn_output.transpose(1, 2).contiguous()
        return attn_output

model = SelfAttention(head_dim, hidden_size, num_attention_heads, context_len=CONTEXT_LEN)

input_embedding = torch.randn(1, SEQ_LEN, hidden_size)
attention_mask = torch.randn(1, 1, SEQ_LEN, CONTEXT_LEN)

with torch.no_grad():
    traced_model = torch.jit.trace(
        model,
        (input_embedding, attention_mask),
        check_trace=True   
    )

inputs = [
    ct.TensorType(name="input_embedding", shape=input_embedding.shape, dtype=np.float16),
    ct.TensorType(name="attention_mask", shape=attention_mask.shape, dtype=np.float16),
]
states = [
    ct.StateType(
        name="k_cache", 
        wrapped_type=ct.TensorType(
            shape=(num_attention_heads, CONTEXT_LEN, head_dim), 
            dtype=np.float16
        )
    ),
    ct.StateType(
        name="v_cache", 
        wrapped_type=ct.TensorType(
            shape=(num_attention_heads, CONTEXT_LEN, head_dim), 
            dtype=np.float16
        )
    ),
]
outputs = [
    ct.TensorType(name="output", dtype=np.float16),
]

converted_model = ct.convert(
    traced_model,
    inputs=inputs,
    outputs=outputs,
    states=states,
    compute_units=ct.ComputeUnit.CPU_AND_NE,
    minimum_deployment_target=ct.target.macOS15,
)
```

My current observation is:
- If I set ComputeUnit as CPU, GPU, it works;
- When I set as CPU_AND_NE, I get error: `You will not be able to run predict() on this Core ML model. Underlying exception message was: {
    NSLocalizedDescription = "Failed to build the model execution plan using a model architecture file '/private/var/folders/qq/6bnk6mx15qndgvplwz2v_y1c0000gn/T/tmpw_5dbl8m.mlmodelc/model.mil' with error code: -14.";`}

My current env info:
- torch: 2.5.0 
- coremltools: 8.3.0

Do you have any insight about this issue? It seems that the ANE is pretty tricky.

**Comment by alexchen4ai** [comment]
Super thanks for the feedback. I have done:
- Force `float16` on export;
    - `model = model.to(torch.float16)`
- Confirm that tensor size not exceeding 4D anywhere;

The error is still there: `NSLocalizedDescription = "Failed to build the model execution plan using a model architecture file '/private/var/folders/qq/6bnk6mx15qndgvplwz2v_y1c0000gn/T/tmplcn3oplx.mlmodelc/model.mil' with error code: -14."`

Also, I checked the netron, haven't found anything strange.

<img width="2580" height="1878" alt="Image" src="https://github.com/user-attachments/assets/9b27751e-7c56-44da-a98e-67f591c0340e" />

**Comment by alexchen4ai** [comment]
I wish to provide more tricky observation:
1. If I choose to comment the slice_update code:
    ```python
    # self.k_cache[:, :seq_len, :] = key_states.squeeze(0)
    # self.v_cache[:, :seq_len, :] = value_states.squeeze(0)
    ```
    It works fine.
2. If I choose to just slice update some random tensors:
    ```python
     self.k_cache[:, :seq_len, :] = torch.randn_like(self.k_cache[:, :seq_len, :])
     self.v_cache[:, :seq_len, :] = torch.randn_like(self.v_cache[:, :seq_len, :])
     ```
      The code can work without any error on ANE.

I had thought this is some special graph arc that ANE can't handle. But, I have learned from you qwen and llama example, which seems to work fine with similar code flow.

**Comment by Anemll** [comment]
Try this, SEQ 16 and 32 seems OK. 
Conv2d for ANE.
Perhaps slice_update size /tensor is too big for  SEQ>32 ,  could be also 64-byte alignment issue for slice_update
hope it helps

Below works

import torch
import numpy as np
import torch.nn as nn
import coremltools as ct

SEQ_LEN, CONTEXT_LEN = 32, 1024
head_dim, hidden_size, num_attention_heads = 32, 1024, 16

class SelfAttention(nn.Module):
    def __init__(self, head_dim, hidden_size, num_attention_heads, context_len=1024):
        super().__init__()
        self.head_dim = head_dim

        self.q_proj = nn.Conv2d(hidden_size, num_attention_heads * head_dim, kernel_size=(1, 1), bias=False)
        self.k_proj = nn.Conv2d(hidden_size, num_attention_heads * head_dim, kernel_size=(1, 1), bias=False)
        self.v_proj = nn.Conv2d(hidden_size, num_attention_heads * head_dim, kernel_size=(1, 1), bias=False)
        self.register_buffer("k_cache", torch.zeros(1, num_attention_heads, context_len, head_dim))
        self.register_buffer("v_cache", torch.zeros(1, num_attention_heads, context_len, head_dim))

    @torch.no_grad()
    def forward(self, hidden_states, attention_mask):
        seq_len = hidden_states.shape[1]
        # Reshape for Conv2D: (batch, seq_len, hidden_size) -> (batch, hidden_size, seq_len, 1)
        hidden_states_4d = hidden_states.transpose(1, 2).unsqueeze(-1)
        
        # Apply Conv2D projections
        query_states = self.q_proj(hidden_states_4d)  # (batch, num_heads * head_dim, seq_len, 1)
        key_states = self.k_proj(hidden_states_4d)    # (batch, num_heads * head_dim, seq_len, 1)
        value_states = self.v_proj(hidden_states_4d)  # (batch, num_heads * head_dim, seq_len, 1)
        
        # Reshape to attention format: (batch, num_heads, seq_len, head_dim)
        query_states = query_states.squeeze(-1).transpose(1, 2).view(1, seq_len, -1, self.head_dim).transpose(1, 2)
        key_states = key_states.squeeze(-1).transpose(1, 2).view(1, seq_len, -1, self.head_dim).transpose(1, 2)
        value_states = value_states.squeeze(-1).transpose(1, 2).view(1, seq_len, -1, self.head_dim).transpose(1, 2)
        
        self.k_cache[0, :, :seq_len, :] = key_states.squeeze(0)
        self.v_cache[0, :, :seq_len, :] = value_states.squeeze(0)
        curr_key_states = self.k_cache
        curr_value_states = self.v_cache

        attn_weights = torch.matmul(query_states, curr_key_states.transpose(2, 3))
        causal_mask = attention_mask[:, :, :, : curr_key_states.shape[-2]]
        attn_weights = attn_weights + causal_mask

        attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float16).to(query_states.dtype)
        attn_output = torch.matmul(attn_weights, curr_value_states)
        attn_output = attn_output.transpose(1, 2).contiguous()
        return attn_output

model = SelfAttention(head_dim, hidden_size, num_attention_heads, context_len=CONTEXT_LEN)

input_embedding = torch.randn(1, SEQ_LEN, hidden_size)
attention_mask = torch.randn(1, 1, SEQ_LEN, CONTEXT_LEN)

with torch.no_grad():
    traced_model = torch.jit.trace(
        model,
        (input_embedding, attention_mask),
        check_trace=True   
    )

inputs = [
    ct.TensorType(name="input_embedding", shape=input_embedding.shape, dtype=np.float16),
    ct.TensorType(name="attention_mask", shape=attention_mask.shape, dtype=np.float16),
]
states = [
    ct.StateType(
        name="k_cache", 
        wrapped_type=ct.TensorType(
            shape=(1, num_attention_heads, CONTEXT_LEN, head_dim), 
            dtype=np.float16
        )
    ),
    ct.StateType(
        name="v_cache", 
        wrapped_type=ct.TensorType(
            shape=(1, num_attention_heads, CONTEXT_LEN, head_dim), 
            dtype=np.float16
        )
    ),
]
outputs = [
    ct.TensorType(name="output", dtype=np.float16),
]

converted_model = ct.convert(
    traced_model,
    inputs=inputs,
    outputs=outputs,
    states=states,
    compute_units=ct.ComputeUnit.CPU_AND_NE,
    compute_precision=ct.precision.FLOAT16,
    minimum_deployment_target=ct.target.iOS18,
)
converted_model.save("/tmp/converted_model.mlpackage")

**Comment by alexchen4ai** [comment]
Thanks, this perfectly solved my problem!
